tests/testthat/type-misc.R

```diff
@@ -0,0 +1,30 @@
+
+test_that("`numeric_version` is a vector (#723)", {
+  x <- utils::packageVersion("rlang")
+  y <- utils::packageVersion("vctrs")
+  z <- c(x, y)
+
+  expect_true(vec_is(x))
+
+  expect_true(vec_equal(x, x))
+  expect_false(vec_equal(x, y))
+  expect_identical(vec_equal(y, z), c(FALSE, TRUE))
+
+  expect_identical(vec_unique(z), z)
+  expect_identical(vec_unique(c(y, z, x)), z[2:1])
+})
+
+test_that("`numeric_version` falls back to base methods", {
+  x <- utils::packageVersion("rlang")
+  y <- utils::packageVersion("vctrs")
+  z <- c(x, y)
+
+  # `z` is a `list-of`-like type but slicing 1 element returns the
```

Comment:
I don't follow this.  Do you mean that `[` returns an atomic type or that `vec_slice()` does?

Reply:
They both do because `vec_slice()` falls back to `[`. To support this natively, without the fallback, we'd need to add a mechanism for wrapper types in vctrs.

Reply:
I don't see how a scalar is different from a vector of length one.

``` r
x <- utils::packageVersion("rlang")
y <- utils::packageVersion("vctrs")
z <- c(x, y)

dput(z)
#> structure(list(c(0L, 4L, 4L), c(0L, 2L, 3L)), class = c("package_version", 
#> "numeric_version"))
dput(z[1:2])
#> structure(list(c(0L, 4L, 4L), c(0L, 2L, 3L)), class = c("package_version", 
#> "numeric_version"))
dput(z[1])
#> structure(list(c(0L, 4L, 4L)), class = c("package_version", "numeric_version"
#> ))
identical(x, x[1])
#> [1] TRUE
identical(x, x[[1]])
#> [1] TRUE
```

<sup>Created on 2020-02-25 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Reply:
oh you're right! I was thrown off by the output of `str()`:

```r
str(z)
#> List of 2
#>  $ :Classes 'package_version', 'numeric_version'  hidden list of 1
#>   ..$ : int [1:4] 0 4 4 9000
#>  $ :Classes 'package_version', 'numeric_version'  hidden list of 1
#>   ..$ : int [1:4] 0 2 99 9005
#>  - attr(*, "class")= chr [1:2] "package_version" "numeric_version"

str(z[1])
#> Classes 'package_version', 'numeric_version'  hidden list of 1
#>  $ : int [1:4] 0 4 4 9000
```

Reply:
Agreed that the weirdness is limited to `str()`. Otherwise even a `numeric_version` of size 1 is a list_of-like object.

The weirdness comes from `irregCl`, which is `TRUE` for a `numeric_version` of size 1, but otherwise is `FALSE` for size >1.
https://github.com/wch/r-source/blob/ec6faae1911207f1cb9b2cf259238598dc1ffe17/src/library/utils/R/str.R#L335

So then the output is changed here based on that
https://github.com/wch/r-source/blob/ec6faae1911207f1cb9b2cf259238598dc1ffe17/src/library/utils/R/str.R#L348

Reply:
It's interesting that `str()` has built in support for lists where `[[` returns the same type.